File: ch01/python/openai-gpt-basic/openai-gpt-basic.py
"""
(C) Copyright 2026 Boni Garcia (https://bonigarcia.github.io/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from openai import OpenAI
import re
import time

client = OpenAI()  # OPENAI_API_KEY should be set as an environment variable


def query_model(prompt: str,
                model: str = "gpt-4o-mini",
                max_tokens: int = 1024,
                temperature: float = 0,
                reasoning: str = "low") -> str:
    """Send a user prompt to an OpenAI model and return the text response."""
    params = {
        "model": model,
        "input": prompt,
        "max_output_tokens": max_tokens,
    }
    if is_gpt5_or_above(model):
        params["reasoning"] = {"effort": reasoning}
    else:
        params["temperature"] = temperature

    start = time.perf_counter()
    response = client.responses.create(**params)
    latency = time.perf_counter() - start

    # Log some details about the response
    usage = response.usage
    print(f"\tModel: {response.model}")
    print(f"\tLatency: {latency:.3f} seconds")
    print(f"\tInput tokens: {usage.input_tokens}")
    print(f"\tOutput tokens: {usage.output_tokens}")
    print(f"\tReasoning tokens: {usage.output_tokens_details.reasoning_tokens}")
    print(f"\tTotal tokens: {usage.total_tokens}")

    return response.output_text


def is_gpt5_or_above(model: str) -> bool:
    return re.match(r"gpt-[5-9].*", model, re.IGNORECASE) is not None


if __name__ == "__main__":
    prompt = "How many tokens is your context window?"

    print("=== Basic model ===")
    print("User:", prompt)
    response = query_model(prompt)
    print("AI:", response)

    print("=== Advanced model ===")
    print("User:", prompt)
    response = query_model(prompt, model="gpt-5", reasoning="medium")
    print("AI:", response)

File: ch01/javascript/openai-gpt-basic/openai-gpt-basic.js
/*
(C) Copyright 2026 Boni Garcia (https://bonigarcia.github.io/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import OpenAI from 'openai';
import { performance } from 'perf_hooks';

const client = new OpenAI(); // OPENAI_API_KEY should be set as an environment variable

async function queryModel(userPrompt, model = "gpt-4o-mini", maxTokens = 1024, temperature = 0, reasoning = "low") {
    const params = {
        model: model,
        input: userPrompt,
        max_output_tokens: maxTokens,
    };

    if (isGpt5OrAbove(model)) {
        params.reasoning = { effort: reasoning };
    } else {
        params.temperature = temperature;
    }

    const start = performance.now();
    const response = await client.responses.create(params);
    const latency = (performance.now() - start) / 1000;

    // Log some details about the response
    const usage = response.usage;
    console.log(`\tModel: ${response.model}`);
    console.log(`\tLatency: ${latency.toFixed(3)} seconds`);
    console.log(`\tInput tokens: ${usage.input_tokens}`);
    console.log(`\tOutput tokens: ${usage.output_tokens}`);
    console.log(`\tReasoning tokens: ${usage.output_tokens_details.reasoning_tokens}`);
    console.log(`\tTotal tokens: ${usage.total_tokens}`);

    return response.output_text;
}

function isGpt5OrAbove(model) {
    return /^gpt-[5-9].*/i.test(model);
}

const userPrompt = "How many tokens is your context window?";

console.log("=== Basic model  ===");
console.log("User:", userPrompt);
var response = await queryModel(userPrompt);
console.log("AI:", response);

console.log("=== Advanced model  ===");
console.log("User:", userPrompt);
response = await queryModel(userPrompt, "gpt-5", 1024, 0, "medium");
console.log("AI:", response);